consensus.go

```diff
@@ -0,0 +1,240 @@
+	w "decred.org/dcrwallet/v2/wallet"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/chaincfg/v3"
+)
+
+type Consensus struct {
+	mwRef *MultiWallet
+	ctx   context.Context
+}
+
+func newConsensus(mwRef *MultiWallet) (*Consensus, error) {
+	c := &Consensus{
+		mwRef: mwRef,
+	}
+
+	return c, nil
+}
+
+// GetVoteChoices handles a getvotechoices request by returning configured vote
+// preferences for each agenda of the latest supported stake version for the
+// selected wallet.
+func (c *Consensus) GetVoteChoices(ctx context.Context, hash string, walletID int) (*GetVoteChoicesResult, error) {
+	wallet := c.mwRef.WalletWithID(walletID)
+	wal := wallet.Internal()
+	if wal == nil {
+		return nil, fmt.Errorf("request requires a wallet but wallet has not loaded yet")
+	}
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return nil, fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	version, agendas := getAllAgendas(wal.ChainParams())
+	resp := &GetVoteChoicesResult{
+		Version: version,
+		Choices: make([]VoteChoice, len(agendas)),
+	}
+
+	choices, _, err := wal.AgendaChoices(ctx, ticketHash)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range choices {
+		resp.Choices[i] = VoteChoice{
+			AgendaID:          choices[i].AgendaID,
+			AgendaDescription: agendas[i].Vote.Description,
+			ChoiceID:          choices[i].ChoiceID,
+			ChoiceDescription: "", // Set below
+		}
+		for j := range agendas[i].Vote.Choices {
+			if choices[i].ChoiceID == agendas[i].Vote.Choices[j].Id {
+				resp.Choices[i].ChoiceDescription = agendas[i].Vote.Choices[j].Description
+				break
+			}
+		}
+	}
+
+	return resp, nil
+}
+
+// SetVoteChoice handles a setvotechoice request by modifying the preferred
+// choice for a voting agenda.
+//
+// If a VSP host is configured in the application settings, the voting
+// preferences will also be set with the VSP.
+func (c *Consensus) SetVoteChoice(walletID int, vspPubKey []byte, vspHost, agendaID, choiceID, hash, passphrase string) error {
+	wallet := c.mwRef.WalletWithID(walletID)
+	wal := wallet.Internal()
+	if wal == nil {
+		return fmt.Errorf("request requires a wallet but wallet has not loaded yet")
+	}
+
+	err := wallet.UnlockWallet([]byte(passphrase))
+	if err != nil {
+		return translateError(err)
+	}
+	defer wallet.LockWallet()
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	choice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: choiceID,
+	}
+
+	ctx := wallet.shutdownContext()
+	_, err = wal.SetAgendaChoices(ctx, ticketHash, choice)
+	if err != nil {
+		return err
+	}
+
+	if vspHost == "" {
+		return fmt.Errorf("request requires a vspHost but no vspHost was provided")
+	}
+
+	vspClient, err := wallet.LookupVSP(vspHost)
```

Comment:
Can do `wallet.VSP(host, pubkey)` directly, that will do a lookup first and create a new one if it doesn't already exist.